File: item-list/src/main/java/com/gurunars/item_list/CollectionManager.java
package com.gurunars.item_list;

import com.esotericsoftware.kryo.Kryo;

import org.objenesis.strategy.StdInstantiatorStrategy;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import java8.util.function.Consumer;


class CollectionManager<ItemType extends Item> implements Serializable {

    private Kryo kryo = new Kryo();

    private List<ItemType> items = new ArrayList<>();
    private Set<ItemType> selectedItems = new HashSet<>();

    private final Consumer<List<SelectableItem<ItemType>>> stateChangeHandler;
    private final Runnable selectionChangeListener;

    CollectionManager(
            Consumer<List<SelectableItem<ItemType>>> stateChangeHandler,
            Runnable selectionChangeListener ) {
        this.stateChangeHandler = stateChangeHandler;
        this.selectionChangeListener = selectionChangeListener;
        this.kryo.setInstantiatorStrategy(new Kryo.DefaultInstantiatorStrategy(new StdInstantiatorStrategy()));
    }

    private void changed(List<ItemType> newItems, Set<ItemType> newSelection) {
        this.items = newItems;
        Set<ItemType> filteredSelection = new HashSet<>();
        for (ItemType cursor: newSelection) {
            if (items.contains(cursor)) {
                // This is to replace the item's payload with a new one
                filteredSelection.add(items.get(items.indexOf(cursor)));
            }
        }
        boolean selectionChanged = !selectedItems.equals(filteredSelection);
        selectedItems = filteredSelection;
        List<SelectableItem<ItemType>> selectableItems = new ArrayList<>();
        for (ItemType item: items) {
            selectableItems.add(new SelectableItem<>(item, selectedItems.contains(item)));
        }

        stateChangeHandler.accept(selectableItems);
        if (selectionChanged) {
            selectionChangeListener.run();
        }
    }

    void itemClick(SelectableItem<ItemType> selectableItem) {
        if (selectedItems.size() == 0) {
            return;
        }

        ItemType item = selectableItem.getItem();

        Set<ItemType> newSelectedItems = new HashSet<>(selectedItems);

        if (newSelectedItems.contains(item)) {
            newSelectedItems.remove(item);
        } else {
            newSelectedItems.add(item);
        }
        changed(items, newSelectedItems);
    }

    void itemLongClick(SelectableItem<ItemType> selectableItem) {
        if (selectedItems.size() != 0) {
            return;
        }
        Set<ItemType> newSelectedItems = new HashSet<>(selectedItems);
        newSelectedItems.add(selectableItem.getItem());
        changed(items, newSelectedItems);
    }

    void setItems(List<ItemType> items) {
        changed(kryo.copy(new ArrayList<>(items)), kryo.copy(new HashSet<>(selectedItems)));
    }

    void setSelectedItems(Set<ItemType> selectedItems) {
        changed(kryo.copy(new ArrayList<>(items)), kryo.copy(new HashSet<>(selectedItems)));
    }

    Set<ItemType> getSelectedItems() {
        return kryo.copy(selectedItems);
    }

}

File: item-list/src/main/kotlin/com/gurunars/item_list/CollectionManager.kt
package com.gurunars.item_list

import com.gurunars.databinding.BindableField
import java.io.Serializable


internal class ItemWrapper<out ItemType: Item>(val item: ItemType) {
    override fun equals(other: Any?) =
        other != null &&
        other is ItemWrapper<*> &&
        item.getId() == other.item.getId()
    override fun hashCode() = item.hashCode()
}


internal fun<ItemType: Item> mergeCollections(items: List<ItemType>, selectedItems: Set<ItemType>) : List<SelectableItem<ItemType>> {
    val all = items.map { ItemWrapper(it) }
    val selected = selectedItems.map { ItemWrapper(it) }
    return all.map { SelectableItem(it.item, selected.contains(it)) }
}


internal fun<ItemType: Item> selectionChange(oldSelection: Set<ItemType>, newSelection: Set<ItemType>) {

}


internal class CollectionManager<ItemType : Item>(
        private val items: BindableField<List<ItemType>>,
        private val selectedItems: BindableField<HashSet<ItemType>>) : Serializable {

    private fun changed(newItems: List<ItemType>, newSelection: Set<ItemType>) {

        val filteredSelection = newSelection
                .filter { indexOfItem(items, it) != -1 }
                .map { items[indexOfItem(items, it)] }
                .toHashSet()
        val selectionChanged = selectedItems != filteredSelection
    }

    fun itemClick(selectableItem: SelectableItem<ItemType>) {
        if (selectedItems.isEmpty()) {
            return
        }

        val item = selectableItem.item
        changed(items,
                if (selectedItems.indexOfFirst { item.getId() == it.getId() } == -1)
                    selectedItems + item
                else
                    selectedItems.filterNot { item.getId() == it.getId() }.toHashSet())
    }

    fun itemLongClick(selectableItem: SelectableItem<ItemType>): Boolean {
        if (selectedItems.isEmpty()) {
            changed(items, selectedItems + hashSetOf(selectableItem.item))
        }
        return true
    }

}
